src/Core/Registration/RegistrationModel.cs

```diff
@@ -6,54 +6,81 @@
-        public RegistrationModel(string registrationId, string email, string clientId, string hash, string salt, RegistrationStep registrationStep)
+        public void CompleteStep(object context)
         {
-            RegistrationId = registrationId;
-            Email = email;
-            ClientId = clientId;
-            Hash = hash;
-            Salt = salt;
-            RegistrationStep = registrationStep;
-        }
+            switch (ActiveStep)
+            {
+                case RegistrationStep.InitialInfo:
+                {
+                    if (context is InitialInfoDto ctx)
+                    {
+                        if (ctx.Email != Email)
+                            throw new ArgumentException("Email doesn't match to verified one.");
+                        if (!IsPasswordComplex(ctx.Password))
+                            throw new PasswordIsNotComplexException();
 
-        private string GenerateRegistrationId()
-        {
-            var guid = Guid.NewGuid();
-            var enc = Convert.ToBase64String(guid.ToByteArray());
-            enc = enc.Replace("/", "_");
-            enc = enc.Replace("+", "-");
-            return enc.Substring(0, 22);
+                        ClientId = ctx.ClientId;
+                        this.SetPassword(ctx.Password);
+
+                        break;
+                    }
+
+                    throw new InvalidRegistrationStepContext(ActiveStep);
+                }
+                case RegistrationStep.AccountInformation:
+                {
+                    if (context is AccountInfoDto ctx)
+                    {
+                        if (!IsPhoneNumberFormatCorrect(ctx.PhoneNumber))
+                            throw new InvalidPhoneNumberFormatException(ctx.PhoneNumber);
+
+                        FirstName = ctx.FirstName;
+                        LastName = ctx.LastName;
+                        CountryCodeIso2 = ctx.CountryCodeIso2;
+                        PhoneNumber = ctx.PhoneNumber;
+
+                        break;
+                    }
+
+                    throw new InvalidRegistrationStepContext(ActiveStep);
+                }
+                default:
+                    throw new NotImplementedException();
+            }
+
+            ActiveStep += 1;
         }
 
-        public void SetInitialInfo(InitialInfoDto initialInfoDto)
+        public static bool IsPhoneNumberFormatCorrect(string phoneNumber)
         {
-            if (initialInfoDto.Email != Email)
-                throw new ArgumentException("Email doesn't match to verified one.");
-            if (!IsPasswordComplex(initialInfoDto.Password))
-                throw new PasswordIsNotComplexException();
-
-            ClientId = initialInfoDto.ClientId;
-            this.SetPassword(initialInfoDto.Password);
-            RegistrationStep = RegistrationStep.AccountInformation;
+            return phoneNumber.PreparePhoneNum().ToE164Number() != null;
```

Comment:
Better to add ? null check
```suggestion
            return phoneNumber?.PreparePhoneNum()?.ToE164Number() != null;
```

Reply:
Fixed